### PR TITLE
Use batch GCS operations during FileSink write finalization

### PR DESCRIPTION
**Pull request by charlesccychen**

### CODE REVIEW

**Comment by charlesccychen** [comment]
R: @chamikaramj

File: sdks/python/apache_beam/io/fileio.py [hunk 1]
```diff
@@ -483,13 +480,49 @@ def rename(src, dst):
...
 
   @staticmethod
+  def rename_batch(src_dest_pairs):
+    # Prepare batches and directly execute local renames.
+    gcs_batches = []
+    gcs_current_batch = []
+    exceptions = []
+    for src, dest in src_dest_pairs:
+      if src.startswith('gs://'):
+        assert dest.startswith('gs://'), dest
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/fileio.py]: Done.

File: sdks/python/apache_beam/io/fileio.py [hunk 2]
```diff
@@ -483,13 +480,49 @@ def rename(src, dst):
...
+      else:
+        try:
+          ChannelFactory.rename(src, dest)
+        except Exception as e:  # pylint: disable=broad-except
+          exceptions.append((src, dest, e))
+    if gcs_current_batch:
+      gcs_batches.append(gcs_current_batch)
+
+    # Execute GCS renames if any and return exceptions.
+    for batch in gcs_batches:
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/fileio.py]: Done.

File: sdks/python/apache_beam/io/fileio.py [hunk 3]
```diff
@@ -815,55 +837,63 @@ def finalize_write(self, init_result, writer_results):
...
 
-    for final_name, err in rename_results:
-      if err:
-        logging.warning('Error when processing rename_results: %s', err)
-        raise err
-      else:
-        yield final_name
+    for exceptions in exception_batches:
+      if exceptions:
+        for e in exceptions:
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/fileio.py]: Done.

File: sdks/python/apache_beam/io/gcsio_test.py [hunk 4]
```diff
@@ -251,6 +281,36 @@ def test_delete(self):
...
+  def test_delete_batch(self, *unused_args):
+    gcsio.BatchApiRequest = FakeBatchApiRequest
+    file_name_pattern = 'gs://gcsio-test/delete_me_%d'
+    file_size = 1024
+    num_files = 10
+
+    # Test deletion of non-existent files.
+    result = self.gcs.delete_batch(
+        [file_name_pattern % i for i in range(num_files)])
+    for i, (file_name, exception) in enumerate(result):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/gcsio_test.py]: Done.

File: sdks/python/apache_beam/io/gcsio_test.py [hunk 5]
```diff
@@ -271,6 +331,43 @@ def test_copy(self):
...
+    from_name_pattern = 'gs://gcsio-test/copy_me_%d'
+    to_name_pattern = 'gs://gcsio-test/destination_%d'
+    file_size = 1024
+    num_files = 10
+
+    # Test copy of non-existent files.
+    result = self.gcs.copy_batch(
+        [(from_name_pattern % i, to_name_pattern % i)
+         for i in range(num_files)])
+    for i, (src, dest, exception) in enumerate(result):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/io/gcsio_test.py]: Done.

**Comment by charlesccychen** [comment]
Thanks, PTAL.

**Comment by charlesccychen** [comment]
Jenkins failure is unrelated.

**Comment by chamikaramj** [comment]
LGTM

**Comment by charlesccychen** [comment]
Thanks Cham!

R: @robertwb

**Comment by robertwb** [comment]
This can now be closed.

**Pull request closed without merging**